src/librustc_mir/monomorphize/collector.rs

```diff
@@ -727,6 +734,10 @@ fn visit_fn_use<'tcx>(
             ty::Instance::resolve_for_fn_ptr
         };
         let instance = resolver(tcx, ty::ParamEnv::reveal_all(), def_id, substs).unwrap();
+
+        let mut resolutions = tcx.call_resolution_map.borrow_mut();
+        resolutions.as_mut().unwrap().insert((def_id, substs), instance);
```

Comment:
This is not correct. Doing a cast to an fnptr can give a different result than a direct call, while both defid and substs are the same.

Reply:
Would it be correct if the hashmap were keyed `(def_id, substs, caller_pos)` where `caller_pos` identifies the position of the call terminator in question?

Reply:
Yes, but only when you don't codegen non generic functions. When codegenning generic functions, you will really need to monomorphize the substs and then use `Instance::resolve*` in the codegen backend.

Reply:
Yes, that's what I feared. We may have to make some fairly substantial changes to our system in light of this. Thanks again for this discussion.

By the way, between us we could probably implement a missing part of the rustc guide here:
https://rust-lang.github.io/rustc-guide/mir/index.html?highlight=instance#mir-data-types

> The main MIR data type is Mir. It contains the data for a single function (along with sub-instances of Mir for "promoted constants", but you can read about those below).

And the link to `below` is "to be written".

So it seems that the type parameters become "promoted constants" during monomorphisation.

(I also think the main MIR data type is now `Body` not `Mir`)

Reply:
> So it seems that the type parameters become "promoted constants" during monomorphisation.

What do you mean?

> (I also think the main MIR data type is now Body not Mir)

`Mir` has been renamed to `Body` recently.

Reply:
> What do you mean?

The wording in the compiler guide talks in these terms, which surprised me a little:

> The main MIR data type is Mir. It contains the data for a single function (along with sub-instances of Mir for "promoted constants"

It's talking about making a generic type parameter into a concrete one, I think? Through the act of promotion.

Reply:
It is actually talking about turning

```rust
fn abc() {
    let a: &'static u8 = 1 + 2;
}
```

into

```rust
fn abc() {
    static PROMOTED0: u8 = 1 + 2;
    let a: &'static u8 = &PROMOTED0;
}
```

Reply:
Oh, I see. We could probably improve the wording there too.